vochain/account.go

```diff
@@ -282,3 +284,124 @@ func SetAccountInfoTxCheck(vtx *models.Tx, txBytes, signature []byte, state *Sta
 }
+
+func MintTokensTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (common.Address, uint64, error) {
+	tx := vtx.GetMintTokens()
+	// check signature available
+	if signature == nil || tx == nil || txBytes == nil {
+		return common.Address{}, 0, fmt.Errorf("missing signature and/or transaction")
+	}
+	// get address from signature
+	sigAddress, err := ethereum.AddrFromSignature(txBytes, signature)
+	if err != nil {
+		return common.Address{}, 0, err
+	}
+	// get treasurer
+	treasurer, err := state.Treasurer(true)
+	if err != nil {
+		return common.Address{}, 0, err
+	}
+	// check signature recovered address
+	tAddr := common.BytesToAddress(treasurer.Address)
+	if tAddr != sigAddress {
+		return common.Address{}, 0, fmt.Errorf("address recovered not treasurer: expected %s got %s", treasurer.String(), sigAddress.String())
+	}
+	// check nonce
+	if tx.Nonce != treasurer.Nonce {
+		return common.Address{}, 0, fmt.Errorf("invalid nonce %d, expected: %d", tx.Nonce, treasurer.Nonce+1)
+	}
+	// check to
+	to := common.BytesToAddress(tx.To)
+	if to.String() == types.EthereumZeroAddress || to.String() == tAddr.String() {
+		return common.Address{}, 0, fmt.Errorf("invalid address to mint")
+	}
+	// check value
+	if tx.Value <= 0 {
+		return common.Address{}, 0, fmt.Errorf("invalid value")
+	}
+	return to, tx.Value, nil
+}
+
+func SetAccountDelegateTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (common.Address, common.Address, error) {
+	tx := vtx.GetSetAccountDelegateTx()
+	// check signature available
+	if signature == nil || tx == nil || txBytes == nil {
+		return common.Address{}, common.Address{}, fmt.Errorf("missing signature and/or transaction")
+	}
+	// get address from signature
+	sigAddress, err := ethereum.AddrFromSignature(txBytes, signature)
+	if err != nil {
+		return common.Address{}, common.Address{}, err
+	}
+	// check nonce
+	acc, err := state.GetAccount(sigAddress, false)
+	if err != nil {
+		return common.Address{}, common.Address{}, fmt.Errorf("cannot get account info: %v", err)
+	}
+	if tx.Nonce != acc.Nonce {
+		return common.Address{}, common.Address{}, fmt.Errorf("invalid nonce, expected %d got %d", acc.Nonce, tx.Nonce)
+	}
+	// check delegate
+	delAcc := common.BytesToAddress(tx.Delegate)
+	if delAcc.String() == types.EthereumZeroAddress {
+		return common.Address{}, common.Address{}, fmt.Errorf("invalid delegate address")
+	}
+	switch tx.Txtype {
+	case models.TxType_ADD_DELEGATE_FOR_ACCOUNT:
+		for i := 0; i < len(acc.DelegateAddrs); i++ {
+			delegateToCmp := common.BytesToAddress(acc.DelegateAddrs[i])
+			if delegateToCmp == delAcc {
+				return common.Address{}, common.Address{}, fmt.Errorf("delegate already added")
+			}
+		}
+		return sigAddress, delAcc, nil
+	case models.TxType_DEL_DELEGATE_FOR_ACCOUNT:
+		f := false
+		for i := 0; i < len(acc.DelegateAddrs); i++ {
+			delegateToCmp := common.BytesToAddress(acc.DelegateAddrs[i])
+			if delegateToCmp == delAcc {
+				f = true
```

Comment:
just put the return here `return sigAddress, delAcc, nil` remove `f`and return by default `return common.Address{}, common.Address{}, fmt.Errorf("cannot remove a non existent delegate")`